File: graphics/Terminal.cs
using System;

namespace TerminalEngine.Graphics;

class Terminal
{
	private int _prevWidth;
	private int _prevHeight;
	private int _terminalWidth;
	private int _terminalHeight;

	public readonly int Width;
	public readonly int Height; 
	public int StartX;
	public int StartY;

	public string borderColor = $"\x1b[38;5;15m\x1b[48;5;70m";
	public string backgroundColor = "\x1b[38;5;16m\x1b[48;5;16m";
	public string screenColor = "\x1b[39m\x1b[49m";

	public char backgroundChar = '#';
	public char borderSidesChar = '|';
	public char borderHorizontalChar = '-';
	public char screenChar = ' ';

	public char cornerTopLeft = '┌';
	public char cornerTopRight = '┐';
	public char cornerBottomLeft = '└';
	public char cornerBottomRight = '┘';

	public Terminal()
	{
		_terminalWidth = Console.WindowWidth;
		_terminalHeight = Console.WindowHeight;
		_prevWidth = _terminalWidth;
		_prevHeight = _terminalHeight;

		Width = _terminalWidth;
		Height = _terminalHeight;
	}
	public Terminal(int width, int height)
	{
		_terminalWidth = Console.WindowWidth;
		_terminalHeight = Console.WindowHeight;
		_prevWidth = _terminalWidth;
		_prevHeight = _terminalHeight;

		Width = width;
		Height = height;
	}

	public void Update()
	{
		_prevWidth = _terminalWidth;
		_prevHeight = _terminalHeight;
		_terminalWidth = Console.WindowWidth;
		_terminalHeight = Console.WindowHeight;
		if (_prevWidth != _terminalWidth || _prevHeight != _terminalHeight)
		{
			DrawBorder();
		}
	}

	public void DrawBorder()
	{
		Console.Write("\x1b[H");
		if (_terminalWidth < Width || _terminalHeight < Height)
		{
			Console.WriteLine("console too small");
		} else
		{
			int diffX = _terminalWidth - Width;
			int diffY = _terminalHeight - Height;

			int topMargin = diffY / 2;
			StartY = topMargin + 1;
			int bottomMargin = diffY - topMargin;
			bottomMargin = _terminalHeight - bottomMargin;
			topMargin -= 1;

			int leftMargin = diffX / 2;
			StartX = leftMargin + 1;
			int rightMargin = diffX - leftMargin;
			rightMargin = _terminalWidth - rightMargin;
			leftMargin -= 1;

			//Console.WriteLine($"width {_terminalWidth} height {_terminalHeight}");
			//Console.WriteLine($"topMargin {topMargin} bottomMargin {bottomMargin}");
			//Console.WriteLine($"leftMargin {leftMargin} rightMargin {rightMargin}");

			for (int i = 0; i<_terminalHeight; i++)
			{
				for (int n = 0; n<_terminalWidth; n++)
				{
					if (i < topMargin || i > bottomMargin)
					{
						if (n == 0)
						{
							Console.Write($"{backgroundColor}{backgroundChar}");
						} else if (n == _terminalWidth - 1)
						{
							if (i != _terminalHeight - 1)
							{
								Console.WriteLine($"{backgroundChar}");
							}else
							{
								Console.Write($"{backgroundChar}");
							}
						} else
						{
							Console.Write($"{backgroundChar}");
						}
					}else if (i == topMargin || i == bottomMargin)
					{
						if (n < leftMargin || n > rightMargin)
						{
							if (n == 0)
							{
								Console.Write($"{backgroundColor}{backgroundChar}");
							} else if (n == _terminalWidth - 1)
							{
								Console.WriteLine($"{backgroundChar}");
							} else
							{
								Console.Write($"{backgroundChar}");
							}
						} else if (n == leftMargin)
						{
							if (i == topMargin)
							{
								Console.Write($"{borderColor}{cornerTopLeft}");
							}else
							{
								Console.Write($"{borderColor}{cornerBottomLeft}");
							}
						}else if (n == rightMargin)
						{
							if (i == topMargin)
							{
								Console.Write($"{cornerTopRight}{backgroundColor}");
							}else
							{
								Console.Write($"{cornerBottomRight}{backgroundColor}");
							}
						} else
						{
							Console.Write($"{borderHorizontalChar}");
						}
					} else
					{
						if (n < leftMargin || n > rightMargin)
						{
							if (n == 0)
							{
								Console.Write($"{backgroundColor}{backgroundChar}");
							} else if (n == _terminalWidth)
							{
								Console.WriteLine($"{backgroundChar}");
							} else
							{
								Console.Write($"{backgroundChar}");
							}
						} else if (n == leftMargin || n == rightMargin)
						{
							Console.Write($"{borderColor}|{backgroundColor}");
						}else if (n == leftMargin + 1)
						{
							Console.Write($"{screenColor}{screenChar}");
						}else
						{
							Console.Write($"{screenChar}");
						}
					}
				}
			}
		}
	}
}

File: graphics/Terminal.py
import os

class Terminal:
    def __init__(self, width, height):
        self.__terminal_size = os.get_terminal_size()
        self.__prev_terminal_size = self.__terminal_size

        self.Width = width
        self.Height = height
        self.Start_x = (self.__terminal_size.columns - self.Width) / 2 + 1
        self.Start_y = (self.__terminal_size.lines - self.Height) / 2 + 1

        self.border_color = "\x1b[38;5;15m\x1b[48;5;70m"
        self.background_color = "\x1b[38;5;16m\x1b[48;5;16m"
        self.screen_color = "\033[0m"

        self.background_char = '#'
        self.border_side_char = '|'
        self.border_horizontal_char = '-'
        self.screen_char = ' '

        self.corner_top_left = '┌'
        self.corner_top_right = '┐'
        self.corner_bottom_left = '└'
        self.corner_bottom_right = '┘'

    def Update(self):
        self.__prev_terminal_size = self.__terminal_size
        self.__terminal_size = os.get_terminal_size()
        if self.__prev_terminal_size != self.__terminal_size:
            self.DrawBorder()

    def DrawBorder(self):
        print("\033[H",end='')
        if self.__terminal_size.columns < self.Width or self.__terminal_size.lines < self.Height:
            print("console too small")
        else:
            diff_x = self.__terminal_size.columns - self.Width
            diff_y = self.__terminal_size.lines - self.Height

            top_margin = diff_y // 2
            Start_y = top_margin + 1
            bottom_margin = diff_y - top_margin
            bottom_margin = self.__terminal_size.lines - bottom_margin
            top_margin -= 1

            left_margin = diff_x // 2
            Start_x = left_margin + 1
            right_margin = diff_x - left_margin
            right_margin = self.__terminal_size.columns - right_margin
            left_margin -= 1

            #print(f"width {self.__terminal_size.columns}, height {self.__terminal_size.lines}")
            #print(f"top {top_margin}, bottom {bottom_margin}")
            #print(f"left {left_margin}, right {right_margin}")

            for i in range(self.__terminal_size.lines):
                for n in range(self.__terminal_size.columns):
                    if i < top_margin or i > bottom_margin:
                        if n == 0:
                            print(f"{self.background_color}{self.background_char}",end="")
                        elif n == self.__terminal_size.columns - 1:
                            if i != self.__terminal_size.lines - 1:
                                print(f"{self.background_char}")
                            else:
                                print(f"{self.background_char}",end="")
                        else:
                            print(f"{self.background_char}",end="")
                    elif i == top_margin or i == bottom_margin:
                        if n < left_margin or n > right_margin:
                            if n == 0:
                                print(f"{self.background_color}{self.background_char}",end="")
                            elif n == self.__terminal_size.columns - 1:
                                print(f"{self.background_char}")
                            else:
                                print(f"{self.background_char}",end="")
                        elif n == left_margin:
                            if i == top_margin:
                                print(f"{self.border_color}{self.corner_top_left}",end="")
                            else:
                                print(f"{self.border_color}{self.corner_bottom_left}",end="")
                        elif n == right_margin:
                            if i == top_margin:
                                print(f"{self.corner_top_right}{self.background_color}",end="")
                            else:
                                print(f"{self.corner_bottom_right}{self.background_color}",end="")
                        else:
                            print(f"{self.border_horizontal_char}",end="")
                    else:
                        if n < left_margin or n > right_margin:
                            if n == 0:
                                print(f"{self.background_color}{self.background_char}",end="")
                            elif n == self.__terminal_size.columns:
                                print(f"{self.background_char}")
                            else:
                                print(f"{self.background_char}",end="")
                        elif n == left_margin or n == right_margin:
                            print(f"{self.border_color}|{self.background_color}",end="")
                        elif n == left_margin + 1:
                            print(f"{self.screen_color}{self.screen_char}",end="")
                        else:
                            print(f"{self.screen_char}",end="")
